miri-script/src/commands.rs

```diff
@@ -0,0 +1,711 @@
+        check_package(
+            &miri_manifest.to_string_lossy().to_owned(),
+            &toolchain,
+            &self.cargo_extra_flags,
+            true,
+            &args,
+        )?;
+        check_package(
+            &cargo_miri_manifest.to_string_lossy().to_owned(),
+            &toolchain,
+            &self.cargo_extra_flags,
+            false,
+            &args,
+        )?;
+        Ok(())
+    }
+
+    /// Returns a toolchain override for use in command line (`+{toolchain_name}`).
+    fn get_toolchain_override(&self) -> String {
+        format!("+{}", &self.active_toolchain)
+    }
+    fn test(&mut self, flags: &[String]) -> Result<()> {
+        let bless = matches!(self.command, Subcommands::Bless { .. });
+        let miri_manifest = path!(self.miri_dir / "Cargo.toml");
+        let miri_manifest = miri_manifest.to_string_lossy().to_owned();
+        let args = flags.iter().map(String::as_str).collect::<Vec<_>>();
+        let toolchain = self.get_toolchain_override();
+        // First build and get a sysroot.
+        Self::build_package(&miri_manifest, &toolchain, &self.cargo_extra_flags, &[])?;
+        self.find_sysroot();
+        if bless {
+            self.extra_flags.insert("MIRI_BLESS".to_owned(), "Gesundheit".to_owned());
+        }
+        // Then test, and let caller control flags.
+        // Only in root project as `cargo-miri` has no tests.
+        cmd(
+            "cargo",
+            chain!(
+                [&toolchain, "test", &self.cargo_extra_flags, "--manifest-path", &miri_manifest],
+                args
+            ),
+        )
+        .run()?;
+        Ok(())
+    }
+
+    fn run_(&mut self, flags: &[String]) -> Result<()> {
+        use itertools::Itertools;
+        let target = flags.iter().tuple_windows().find(|(first, _)| first.as_str() == "--target");
+        if let Some(target) = target {
+            self.extra_flags.insert("MIRI_TEST_TARGET".to_owned(), target.1.clone());
+        } else if let Some(var) =
+            std::env::var("MIRI_TEST_TARGET").ok().filter(|target| !target.is_empty())
+        {
+            // Make sure miri actually uses this target.
+            self.extra_flags
+                .entry("MIRIFLAGS".to_owned())
+                .or_default()
+                .push_str(&format!(" --target {var}"));
+        }
+        // First build and get a sysroot.
+        let miri_manifest = path!(self.miri_dir / "Cargo.toml");
+        let miri_manifest = miri_manifest.to_string_lossy();
+        let toolchain = self.get_toolchain_override();
+        Self::build_package(&miri_manifest, &toolchain, &self.cargo_extra_flags, &[])?;
+        self.find_sysroot();
+        // Then run the actual command.
+        let miri_flags = self.extra_flags.get("MIRIFLAGS").cloned();
+        let miri_flags = miri_flags.as_deref();
+        let args = flags.iter().map(String::as_str).collect::<Vec<_>>();
```

Comment:
Does `chain!` really force us to use `Vec<&str>`? Would be much more ergonomic if it worked more like `Command::args` where we can give any iterator with element type `AsRef<OsStr>`, then we would have less conversion noise here

Reply:
Chain required Item types of chained iterators to match, so there was a need for conversion. I guess it's redundant now?